File: src/main/java/pl/training/patterns/behavioral/observer/ServerEvent.java
package pl.training.patterns.behavioral.observer;

class ServerEvent {
    private String payload;

    ServerEvent(final String payload) {
        this.payload = payload;
    }


    public static class ServerEventBuilder {
        private String payload;

        ServerEventBuilder() {
        }

        /**
         * @return {@code this}.
         */
        public ServerEvent.ServerEventBuilder payload(final String payload) {
            this.payload = payload;
            return this;
        }

        public ServerEvent build() {
            return new ServerEvent(this.payload);
        }

        @Override
        public String toString() {
            return "ServerEvent.ServerEventBuilder(payload=" + this.payload + ")";
        }
    }

    public static ServerEvent.ServerEventBuilder builder() {
        return new ServerEvent.ServerEventBuilder();
    }

    public String getPayload() {
        return this.payload;
    }
}

File: csharp/src/Training.Patterns/Behavioral/Observer/ServerEvent.cs
namespace Training.Patterns.Behavioral.Observer;

internal class ServerEvent
{
    public string? Payload { get; }

    internal ServerEvent(string? payload)
    {
        Payload = payload;
    }

    public class ServerEventBuilder
    {
        private string? payload;

        internal ServerEventBuilder()
        {
        }

        /// <returns><c>this</c>.</returns>
        public ServerEventBuilder Payload(string? payload)
        {
            this.payload = payload;
            return this;
        }

        public ServerEvent Build() => new(payload);

        public override string ToString() => "ServerEvent.ServerEventBuilder(payload=" + (payload ?? "null") + ")";
    }

    public static ServerEventBuilder Builder() => new();
}
